### PR TITLE
[CARBONDATA-1364] Added the blocklet info to index file and make the datamap distributable with job

### PR DESCRIPTION
**Pull request by ravipesala**
In this PR following tasks are completed.
1. Added the blocklet info to the carbonindex file, so datamap not required to read each carbondata file footer to the blocklet information. This makes the datamap loading faster.

2. Made the data map distributable and added the spark job. So datamap pruning could happen distributable and pruned blocklet list would be sent to driver.

This PR cannot compile as carbondata format changes are present.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/504/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3094/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3218/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/622/

**Comment by CarbonDataQA** [comment]
SDV Build Failed with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/38/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3391/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/794/

**Comment by CarbonDataQA** [comment]
SDV Build Failed with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/121/

**Comment by CarbonDataQA** [comment]
SDV Build Failed with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/122/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/795/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3393/

File: core/src/main/java/org/apache/carbondata/core/indexstore/DataMapDistributable.java [hunk 1]
```diff
@@ -53,4 +63,27 @@ public void setDataMapName(String dataMapName) {
...
+
+  public DataMapType getDataMapType() {
+    return dataMapType;
+  }
+
+  public void setDataMapType(DataMapType dataMapType) {
+    this.dataMapType = dataMapType;
+  }
+
+  @Override public String[] getLocations() throws IOException {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/indexstore/DataMapDistributable.java]: move @Override to previous line

File: core/src/main/java/org/apache/carbondata/core/indexstore/DataMapStoreManager.java [hunk 2]
```diff
@@ -79,15 +79,15 @@ private TableDataMap createTableDataMap(AbsoluteTableIdentifier identifier,
...
+    TableDataMap dataMap = getTableDataMap(dataMapName, tableDataMaps);
     if (dataMap != null) {
       throw new RuntimeException("Already datamap exists in that path with type " + mapType);
     }
 
     try {
       DataMapFactory dataMapFactory = mapType.getClassObject().newInstance();
       dataMapFactory.init(identifier, dataMapName);
-      dataMap = new TableDataMap(identifier, dataMapName, dataMapFactory);
+      dataMap = new TableDataMap(identifier, dataMapName, mapType, dataMapFactory);
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/indexstore/DataMapStoreManager.java]: rename `mapType` to `dataMapType`

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMap.java [hunk 3]
```diff
@@ -88,11 +87,14 @@
 
   private int[] columnCardinality;
 
+  private String indexFilePath;
+
   @Override public DataMapWriter getWriter() {
     return null;
   }
 
   @Override public void init(String path) {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMap.java]: suggest change `path` to `filePath`

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapDistributable.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+package org.apache.carbondata.core.indexstore.blockletindex;
+
+import org.apache.carbondata.core.indexstore.DataMapDistributable;
+
+/**
+ * Blocklet datamap distributable
+ */
+public class BlockletDataMapDistributable extends DataMapDistributable {
+
+  private String indexFileName;
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapDistributable.java]: path or name?

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/DistributableDataMapFormat.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+      Object toObject = ObjectSerializationUtil.convertStringToObject(filterString);
+      return (FilterResolverIntf) toObject;
+    }
+    return null;
+  }
+
+  @Override public List<InputSplit> getSplits(JobContext job) throws IOException {
+    TableDataMap dataMap =
+        DataMapStoreManager.getInstance().getDataMap(identifier, dataMapName, dataMapType);
+    List<DataMapDistributable> distributables = dataMap.toDistributable(validSegments);
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/DistributableDataMapFormat.java]: Can we first filter on the segment based on minmax ? If segments are more, even map reduce will take times

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonScanRDD.scala [hunk 6]
```diff
@@ -249,6 +249,8 @@ class CarbonScanRDD(
 
   private def prepareInputFormatForDriver(conf: Configuration): CarbonTableInputFormat[Object] = {
     CarbonTableInputFormat.setTableInfo(conf, tableInfo)
+    // to make the datamap pruning distributable
+    // CarbonTableInputFormat.setDataMapJob(conf, new SparkDataMapJob)
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonScanRDD.scala]: Is this required?

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    dataMapFormat: DistributableDataMapFormat,
+    resolverIntf: FilterResolverIntf)
+  extends CarbonRDD[(Blocklet)](sc, Nil) {
+
+  private val jobTrackerId: String = {
+    val formatter = new SimpleDateFormat("yyyyMMddHHmm")
+    formatter.format(new Date())
+  }
+
+  override def internalCompute(split: Partition,
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala]: please add some comment

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+import org.apache.carbondata.core.indexstore.Blocklet
+import org.apache.carbondata.core.scan.filter.resolver.FilterResolverIntf
+import org.apache.carbondata.hadoop.api.{DataMapJob, DistributableDataMapFormat}
+
+/**
+ * Spark job to execute datamap job and prune all the datamaps distributable
+ */
+class SparkDataMapJob extends DataMapJob {
+
+
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala]: remove this

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  }
+}
+
+class SparkDataMapPartition(rddId: Int, idx: Int, val inputSplit: InputSplit) extends Partition {
+  override def index: Int = idx
+
+  override def hashCode(): Int = 41 * (41 + rddId) + idx
+}
+
+class SparkDataMapRDD(sc: SparkContext,
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala]: please add some comment. suggest change to `DataMapScanRDD`

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/DistributableDataMapFormat.java (hunk 5 above)
Comment (ravipesala) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/DistributableDataMapFormat.java]: you mean first prune using default datamap(blockletdatamap) in the driver side and then distribute the job. But what if blockletdatamap needs to be distributed? I think this intelligence need to be add at the time of choosing the datamap/datamaps and do the datamap chaining. May we can do this after we add another datamap(secondary index)

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonScanRDD.scala (hunk 6 above)
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/CarbonScanRDD.scala]: ok, removed

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala (hunk 8 above)
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala]: ok

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala (hunk 9 above)
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/rdd/SparkDataMapJob.scala]: ok

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapDistributable.java (hunk 4 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapDistributable.java]: changed

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMap.java (hunk 3 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMap.java]: ok

File: core/src/main/java/org/apache/carbondata/core/indexstore/DataMapStoreManager.java (hunk 2 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/DataMapStoreManager.java]: ok

File: core/src/main/java/org/apache/carbondata/core/indexstore/DataMapDistributable.java (hunk 1 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/indexstore/DataMapDistributable.java]: ok

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-2.1/91/

**Comment by ravipesala** [comment]
SDV Build Failed with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/139/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/carbondata-pr-spark-1.6/743/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://144.76.159.231:8080/job/ApacheCarbonPRBuilder/820/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/3412/

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapDistributable.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.carbondata.core.indexstore.blockletindex;
+
+import org.apache.carbondata.core.indexstore.DataMapDistributable;
+
+/**
+ * Blocklet datamap distributable
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapDistributable.java]: Can you add more description to describe its functionality

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java [hunk 11]
```diff
@@ -84,6 +82,38 @@ public DataMapWriter getDataMapWriter(AbsoluteTableIdentifier identifier, String
     }
   }
 
+  private CarbonFile[] getCarbonIndexFiles(String segmentId) {
+    String path = identifier.getTablePath() + "/Fact/Part0/Segment_" + segmentId;
+    FileFactory.FileType fileType = FileFactory.getFileType(path);
+    CarbonFile carbonFile = FileFactory.getCarbonFile(path, fileType);
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java]: use `public static CarbonFile getCarbonFile(String path, FileType fileType)`

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapDistributable.java (hunk 10 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapDistributable.java]: ok

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java (hunk 11 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java]: You mean use `FileFactory.getCarbonFile(path)` ? I have changed

**Comment by CarbonDataQA** [comment]
SDV Build Failed with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/313/

**Comment by CarbonDataQA** [comment]
SDV Build Failed with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/314/

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/319/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**